### PR TITLE
Alibaba Cloud backend supports CR in VPC. [BA-5928]

### PR DESCRIPTION
**Pull request by ysp0606**

### CODE REVIEW

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsConfiguration.scala [hunk 1]
```diff
@@ -44,6 +44,7 @@ final class BcsConfiguration(val configurationDescriptor: BackendConfigurationDe
   lazy val dockerHashAccessId = DockerConfiguration.dockerHashLookupConfig.as[Option[String]]("alibabacloudcr.auth.access-id")
   lazy val dockerHashAccessKey = DockerConfiguration.dockerHashLookupConfig.as[Option[String]]("alibabacloudcr.auth.access-key")
   lazy val dockerHashSecurityToken = DockerConfiguration.dockerHashLookupConfig.as[Option[String]]("alibabacloudcr.auth.security-token")
+  lazy val dockerHashEndpoint: Option[String] = DockerConfiguration.dockerHashLookupConfig.as[Option[String]]("alibabacloudcr.auth.endpoint")
```
Comment (kshakir) [inline supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsConfiguration.scala]: This is consistent with the previous lines, so no need to change this line. For those looking for a slightly shorter Option...

```scala
lazy val dockerHashEndpoint: Option[String] = DockerConfiguration.dockerHashLookupConfig.getAs[String]("alibabacloudcr.auth.endpoint")
```

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsBackendLifecycleActorFactory.scala [hunk 2]
```diff
@@ -31,7 +31,12 @@ final case class BcsBackendLifecycleActorFactory(val name: String, val configura
   override def dockerHashCredentials(workflowDescriptor: BackendWorkflowDescriptor, initializationData: Option[BackendInitializationData]) = {
     Try(BackendInitializationData.as[BcsBackendInitializationData](initializationData)) match {
       case Success(bcsData) =>
-        List(bcsData.bcsConfiguration.dockerCredentials).flatten
+        {
+          bcsData.bcsConfiguration.dockerHashEndpoint match {
+            case Some(endpoint) => List(bcsData.bcsConfiguration.dockerCredentials, Option(endpoint)).flatten
+            case None => List(bcsData.bcsConfiguration.dockerCredentials).flatten
+          }
+        }
```
[the remark this answers is not in the document]
Comment (ysp0606) [reply supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsBackendLifecycleActorFactory.scala]: removed

File: dockerHashing/src/main/scala/cromwell/docker/registryv2/flows/alibabacloudcr/AlibabaCloudCRRegistry.scala [hunk 3]
```diff
@@ -92,6 +104,16 @@ class AlibabaCloudCRRegistry(config: DockerRegistryConfig) extends DockerRegistr
...
 
+  //cr.cn-beijing.aliyuncs.com  or  cr-vpc.cn-beijing.aliyuncs.com
+  private[alibabacloudcrregistry] def getAliyunEndpointFromContext(context: DockerInfoContext): Option[String] = {
+    context.credentials find {
+      _.isInstanceOf[String]
+    } match {
+      case Some(endpoint: String) => Option(endpoint).filter(isValidAlibabaCloudCREndpoint)
+      case _ => None
+    }
+  }
```
[the remark this answers is not in the document]
Comment (ysp0606) [reply dockerHashing/src/main/scala/cromwell/docker/registryv2/flows/alibabacloudcr/AlibabaCloudCRRegistry.scala]: OK, Thanks!

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsConfiguration.scala (hunk 1 above)
Comment (ysp0606) [reply supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsConfiguration.scala]: Done

**Comment by kshakir** [comment]
Ran commit `34fcb6bf3b9e29557a7d9ac057e8ae07370180b9` on my laptop via `sbt clean && src/ci/bin/testCentaurBcs.sh` and tests passed. Dunno if there was a GitHub outage missing the event, but [the Travis CI test passed](https://travis-ci.com/broadinstitute/cromwell/builds/123676180).

Merging.

**Pull request merged**